extensions/amp-ad-network-doubleclick-impl/0.1/safeframe-host.js

```diff
@@ -0,0 +1,580 @@
+    this.iframe_ = this.baseInstance_.iframe;
+    this.channel = channel;
+    dev().assert(this.iframe_.contentWindow,
+        'Frame contentWindow unavailable.');
+    this.setupGeom_();
+    this.sendMessage_(dict({
+      'message': 'connect',
+      'c': this.channel,
+    }), '');
+  }
+
+  /**
+   * Creates IntersectionObserver instance for this SafeframeAPI instance.
+   * We utilize the existing IntersectionObserver class, by passing in this
+   * class for IO to use instead of SubscriptionApi for sending its update
+   * messages. The method 'send' below is triggered by IO every time that
+   * an update occurs.
+   * @private
+   */
+  setupGeom_() {
+    this.IntersectionObserver_ = new IntersectionObserver(
+        this.baseInstance_, this.baseInstance_.element, false, this, 1000);
+    // This will send a geometry update message, and then start sending
+    // them whenever geometry changes as detected by intersection
+    // observer. Make sure we always send this initial update message.
+    this.IntersectionObserver_.startSendingIntersectionChanges();
+  }
+
+  /**
+   * Every time that the IntersectionObserver needs to send an update, this
+   * method is triggered. Includes page scroll or other visibility change
+   * events.
+   * Handles sending the geometry update message to the
+   * safeframe container, which allows $sf.ext.geom() to work.
+   */
+  send(unusedTrash, changes) {
+    this.sendMessage_({
+      newGeometry: this.formatGeom_(changes['changes'][0]),
+      uid: this.uid,
+    }, SERVICE.GEOMETRY_UPDATE);
+  }
+
+  /**
+   * The IntersectionObserver class is monitoring the amp-ad element,
+   * not the Safeframe. The Safeframe is not necessarily the exact
+   * same size as the amp-ad element, so calculate the width and height
+   * correction between the amp-ad element and the safeframe iframe, and then
+   * modify the change entry to be correct.
+   * @param {!Object} changes
+   * @return {!Object} Corrected intersection change entry.
+   */
+  correctChanges(changes) {
+    const iframeRect = this.iframe_ ? this.iframe_.getBoundingClientRect() :
+      this.creativeSize_;
+    const ampAdRect = this.baseInstance_.element.getBoundingClientRect();
+    const widthCorrection = (ampAdRect.width - iframeRect.width) / 2;
+    const heightCorrection = (ampAdRect.height - iframeRect.height) / 2;
+    changes.boundingClientRect.right -= widthCorrection;
+    changes.boundingClientRect.top += heightCorrection;
+    changes.boundingClientRect.bottom -= heightCorrection;
+    changes.boundingClientRect.left += widthCorrection;
+    return changes;
+  }
+
+  /**
+   * Converts an IntersectionObserver-formatted change message to the
+   * geometry update format expected by GPT Safeframe.
+   * @param {!Object} changes IntersectionObserver formatted
+   *   changes.
+   * @return {string} Safeframe formatted changes.
+   * @private
+   */
+  formatGeom_(changes) {
+    const percInView = (a1, b1, a2, b2) => {
```

Comment:
Can we name this parameters something useful?  You have a field in the function that you happened to name the same as the function?